client/network/src/discovery.rs

```diff
@@ -479,6 +573,170 @@ impl NetworkBehaviour for DiscoveryBehaviour {
 	}
 }
 
+pub struct DiscoveryHandler {
```

Comment:
I would really prefer if ultimately this kind of combinators was in libp2p rather than Substrate.
What about opening a libp2p issue and adding a `TODO` here?